dnssec_keyscan.go

```diff
@@ -166,6 +171,30 @@ func readPrivateKeyECDSA(m map[string]string) (*ecdsa.PrivateKey, error) {
 	return p, nil
 }
 
+func readPrivateKeyED25519(m map[string]string) (ed25519.PrivateKey, error) {
+	var p ed25519.PrivateKey
+	// TODO: validate that the required flags are present
+	for k, v := range m {
+		switch k {
+		case "privatekey":
+			p1, err := fromBase64([]byte(v))
+			if err != nil {
+				return nil, err
+			}
+			if len(p1) != 32 {
+				return nil, ErrPrivKey
+			}
+			_, p, err = ed25519.GenerateKey(bytes.NewReader(p1))
```

Comment:
The interface in the ed25519 pkgs uses GenerateKey to reconstruct a key? There is no e.g. setBytes equiv?

Reply:
@miekg Unfortunately the representation of ed25519 private keys in Golang and RFC 8080 differ. RFC 8080 uses only the seed (or the random private half), while Golang uses `seed || public-key` – in ed25519 the public key is derived from the seed. `ed25519.GenerateKey` uses the first 32-bytes read from `rand` (`bytes.NewReader(p1)`) as the seed/private-half and then derives the public half from this. I'm using `ed25519.GenerateKey` in the same fashion as [`ED25519_keypair_from_seed` from BoringSSL](https://github.com/google/boringssl/blob/f21650709a6f76e829ddcc77fe221c9d6a5c12de/include/openssl/curve25519.h#L94-L102).

I'll completely agree that this isn't particularly obvious and I can add documentation to this. I'm not sure exactly what that documentation should be though.

Reply:
Something like your comment, maybe somewhat condensed?

Reply:
@miekg I've got something worked out now, but I'm not sure whether to draw the parallel to ED25519_keypair_from_seed.

```
// ...
//
// This is used in the same manner as |ED25519_keypair_from_seed|
// from BoringSSL.
```

Thoughts?

Reply:
OK. Not sure about referencing a C impl. in the comments makes sense for a Go library though? Did you push your changes?

Reply:
@miekg That was my thinking too. I've pushed them now, I was just waiting to hear what you thought first.